### PR TITLE
feat: add exception logging function, add tests

### PR DESCRIPTION
**Pull request by sgoral-splunk**
this feature adds exception logging functionality

### CODE REVIEW

File: solnlib/log.py [hunk 1]
```diff
@@ -263,3 +265,25 @@ def events_ingested(
...
     )
+
+
+def log_exception(
+    logger: logging.Logger,
+    e: Exception,
+    full_msg: bool = True,
+    msg_before: str = None,
+    msg_after: str = None,
+    log_level: int = logging.INFO,
```
Comment (artemrys) [inline solnlib/log.py:276]: logging.ERROR might be a better default value for such log?

**Review by artemrys — CHANGES_REQUESTED** [review]
This is what I got during my testing:

```
    def stream_events(self, inputs, event_writer):
        for input_name, input_item in inputs.inputs.items():
            normalized_input_name = input_name.split("/")[-1]
            logger = logger_for_input(normalized_input_name)
            try:
                session_key = self._input_definition.metadata["session_key"]
                log_level = conf_manager.get_log_level(
                    logger=logger,
                    session_key=session_key,
                    app_name=ADDON_NAME,
                    conf_name=f"{ADDON_NAME}_settings",
                )
                logger.setLevel(log_level)
                log.modular_input_start(logger, normalized_input_name)
                data = get_data_from_api(logger, "hello world")
                sourcetype = "dummy-data"
                for line in data:
                    event_writer.write_event(
                        smi.Event(
                            data=json.dumps(line, ensure_ascii=False, default=str),
                            index=input_item.get("index"),
                            sourcetype=sourcetype,
                        )
                    )
                raise ValueError("throwing a random exception out of nowhere")
                log.events_ingested(
                    logger, normalized_input_name, sourcetype, len(data)
                )
                log.modular_input_end(logger, normalized_input_name)
            except Exception as e:
                log.log_exception(
                    logger,
                    e,
                    msg_before="Exception raised while ingesting data for ",
                    log_level=logging.ERROR,
                )
```

image

File: solnlib/log.py [hunk 2]
```diff
@@ -19,6 +19,8 @@
 import logging
 import logging.handlers
 import os.path as op
+import sys
```
[the remark this answers is not in the document]
Comment (sgoral-splunk) [reply solnlib/log.py:22]: Right, maybe we can add flake8 to pre-commit?

**Comment by sgoral-splunk** [comment]
that's odd, because it worked for me every time and it looks like this:
image
I will try to reproduce the error

File: solnlib/log.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (artemrys) [reply solnlib/log.py:22]: Let's do this.

**Review by artemrys — APPROVED** [review]

**Pull request merged**